Quarkus 3.2.2.Final issue with DynamoDbEnhancedAsyncClient and InjectMock (mockk)

### Hi.

### After update Quarkus to 3.2.2.Final I faced weird behavior related to DynamoDbEnhancedAsyncClient.

### Link to github project for easier issue reproduction: https://github.com/DudekJakub/Quarkus_DynamoDB_demo

---

Project is split into 5 branches:

`3.2.0_DYNAMO_OK` -> here everything works as expected, _DynamoDbEnhancedAsyncClient_ works properly.

`3.2.0_DYNAMO_with_InjectMock_OK` -> here I added _Clock_ (java.time.Clock) which is being injected (with **mockk**) in _GreetingResourceTest_. Everything works smoothly, no exceptions.

`3.2.2_DYNAMO_ERROR` -> here _DynamoDbEnhancedAsyncClient_ cannot be used due to "**Ambiguous dependencies for type software.amazon.awssdk.enhanced.dynamodb.DynamoDbEnhancedAsyncClient and qualifiers [@Default]**" (no Clock injection to simplify the case).

`3.2.2_DYNAMO_with_@DefaultBean_OK` -> here the code is the same as in above branch but with one change. I decided to add **@DefaultBean** annotation over bean declaration in _Config_ class -> it works (no Clock injection).

`3.2.2_DYNAMO_with_@DefaultBean_and_InjectMock_ERROR` -> here the code is the same as in above branch but with one change. I decided to add Clock back and inject it (with mockk) in _GreetingResourceTest_ to mock its behavior (please check _Config_ and _GreetingResourceTest_ classes to see how its done) -> _DynamoDbEnhancedAsyncClient_ bean with the mentioned annotation works but now it cannot inject mockked Clock (another exception).

---

`JAVA version: `
openjdk 17.0.7 2023-04-18 LTS
OpenJDK Runtime Environment Corretto-17.0.7.7.1 (build 17.0.7+7-LTS)
OpenJDK 64-Bit Server VM Corretto-17.0.7.7.1 (build 17.0.7+7-LTS, mixed mode, sharing)

`GRADLE version:` 8.1.1

`Quarkus version:` 3.2.0.Final and 3.2.2.Final

`Junit5-mockk version:` 2.0.0

---

You may ignore other branches as they are related to original post I wrote here:
https://github.com/quarkusio/quarkus/issues/34924 - these branches include 3.2.1.Final version as well (same issue).

---

Could anyone provide me some feedback if there is something I do wrong here or why this issue occurs at all?
Thank you very much in advance. 

---

Here are parts of stacktraces from above ERROR branches:

1) `3.2.2_DYNAMO_ERROR`:

> java.lang.RuntimeException: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
> 	[error]: Build step io.quarkus.arc.deployment.ArcProcessor#validate threw an exception: jakarta.enterprise.inject.spi.DeploymentException: jakarta.enterprise.inject.AmbiguousResolutionException: Ambiguous dependencies for type software.amazon.awssdk.enhanced.dynamodb.DynamoDbEnhancedAsyncClient and qualifiers [@Default]
> 	- java member: org.acme.GreetingResource():dynamoClient
> 	- declared on CLASS bean [types=[org.acme.GreetingResource, java.lang.Object], qualifiers=[@Default, @Any], target=org.acme.GreetingResource]
> 	- available beans:
> 		- PRODUCER METHOD bean [types=[software.amazon.awssdk.enhanced.dynamodb.DynamoDbEnhancedAsyncClient, software.amazon.awssdk.enhanced.dynamodb.DynamoDbEnhancedResource, java.lang.Object], qualifiers=[@Default, @Any], target=software.amazon.awssdk.enhanced.dynamodb.DynamoDbEnhancedAsyncClient enhancedClient(software.amazon.awssdk.services.dynamodb.DynamoDbAsyncClient client), declaringBean=org.acme.Config]
> 		- SYNTHETIC bean [types=[software.amazon.awssdk.enhanced.dynamodb.DynamoDbEnhancedAsyncClient, java.lang.Object], qualifiers=[@Default, @Any], target=n/a]

	

2) `3.2.2_DYNAMO_with_@DefaultBean_and_InjectMock_ERROR`:

> Failed to create test instance
> org.junit.jupiter.api.extension.TestInstantiationException: Failed to create test instance
> 	at io.quarkus.test.junit.QuarkusTestExtension.initTestState(QuarkusTestExtension.java:786)
> 	at io.quarkus.test.junit.QuarkusTestExtension.interceptTestClassConstructor(QuarkusTestExtension.java:752)
> 	at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker.lambda$invoke$0(InterceptingExecutableInvoker.java:93)
> 	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:106)
> 	at org.junit.jupiter.api.extension.InvocationInterceptor.interceptTestClassConstructor(InvocationInterceptor.java:73)
> 	at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker.lambda$invoke$0(InterceptingExecutableInvoker.java:93)
> 	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:106)
> 	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.proceed(InvocationInterceptorChain.java:64)
> 	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.chainAndInvoke(InvocationInterceptorChain.java:45)
> 	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.invoke(InvocationInterceptorChain.java:37)
> 	at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker.invoke(InterceptingExecutableInvoker.java:92)
> 	at org.junit.jupiter.engine.execution.InterceptingExecutableInvoker.invoke(InterceptingExecutableInvoker.java:62)
> 	at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.invokeTestClassConstructor(ClassBasedTestDescriptor.java:363)
> 	at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.instantiateTestClass(ClassBasedTestDescriptor.java:310)
> 	at org.junit.jupiter.engine.descriptor.ClassTestDescriptor.instantiateTestClass(ClassTestDescriptor.java:79)
> 	at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.instantiateAndPostProcessTestInstance(ClassBasedTestDescriptor.java:286)
> 	at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.lambda$testInstancesProvider$4(ClassBasedTestDescriptor.java:278)
> 	at java.base/java.util.Optional.orElseGet(Optional.java:364)
> 	at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.lambda$testInstancesProvider$5(ClassBasedTestDescriptor.java:277)
> 	at org.junit.jupiter.engine.execution.TestInstancesProvider.getTestInstances(TestInstancesProvider.java:31)
> 	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$prepare$0(TestMethodTestDescriptor.java:105)
> 	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
> 	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.prepare(TestMethodTestDescriptor.java:104)
> 	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.prepare(TestMethodTestDescriptor.java:68)
> 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$prepare$2(NodeTestTask.java:123)
> 	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
> 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.prepare(NodeTestTask.java:123)
> 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:90)
> 	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
> 	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
> 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
> 	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
> 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
> 	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
> 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
> 	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
> 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
> 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
> 	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
> 	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
> 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
> 	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
> 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
> 	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
> 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
> 	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
> 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
> 	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
> 	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:35)
> 	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
> 	at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:54)
> 	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:147)
> 	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:127)
> 	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:90)
> 	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.lambda$execute$0(EngineExecutionOrchestrator.java:55)
> 	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.withInterceptedStreams(EngineExecutionOrchestrator.java:102)
> 	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:54)
> 	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:114)
> 	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:86)
> 	at org.junit.platform.launcher.core.DefaultLauncherSession$DelegatingLauncher.execute(DefaultLauncherSession.java:86)
> 	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor$CollectAllTestClassesExecutor.processAllTestClasses(JUnitPlatformTestClassProcessor.java:110)
> 	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor$CollectAllTestClassesExecutor.access$000(JUnitPlatformTestClassProcessor.java:90)
> 	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor.stop(JUnitPlatformTestClassProcessor.java:85)
> 	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.stop(SuiteTestClassProcessor.java:62)
> 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
> 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
> 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
> 	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
> 	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
> 	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
> 	at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:33)
> 	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:94)
> 	at jdk.proxy1/jdk.proxy1.$Proxy2.stop(Unknown Source)
> 	at org.gradle.api.internal.tasks.testing.worker.TestWorker$3.run(TestWorker.java:193)
> 	at org.gradle.api.internal.tasks.testing.worker.TestWorker.executeAndMaintainThreadName(TestWorker.java:129)
> 	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:100)
> 	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:60)
> 	at org.gradle.process.internal.worker.child.ActionExecutionWorker.execute(ActionExecutionWorker.java:56)
> 	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:113)
> 	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:65)
> 	at worker.org.gradle.process.internal.worker.GradleWorkerMain.run(GradleWorkerMain.java:69)
> 	at worker.org.gradle.process.internal.worker.GradleWorkerMain.main(GradleWorkerMain.java:74)
> Caused by: java.lang.IllegalStateException: Invalid use of io.quarkiverse.test.junit.mockk.InjectMock - could not determine bean of type: class java.time.Clock. Offending field is clock of test class class org.acme.GreetingResourceTest

--- a/dynamodb-enhanced/deployment/src/main/java/io/quarkus/amazon/dynamodb/enhanced/deployment/DynamodbEnhancedProcessor.java
+++ b/dynamodb-enhanced/deployment/src/main/java/io/quarkus/amazon/dynamodb/enhanced/deployment/DynamodbEnhancedProcessor.java
@@ -137,6 +137,7 @@ public class DynamodbEnhancedProcessor {
             if (syncClientRuntime != null) {
                 syntheticBean.produce(SyntheticBeanBuildItem
                         .configure(DynamoDbEnhancedClient.class)
+                        .defaultBean()
                         .scope(ApplicationScoped.class)
                         .setRuntimeInit()
                         .createWith(recorder.createDynamoDbEnhancedClient(extensions))
@@ -147,6 +148,7 @@ public class DynamodbEnhancedProcessor {
             if (asyncClientRuntime != null) {
                 syntheticBean.produce(SyntheticBeanBuildItem
                         .configure(DynamoDbEnhancedAsyncClient.class)
+                        .defaultBean()
                         .scope(ApplicationScoped.class)
                         .setRuntimeInit()
                         .createWith(recorder.createDynamoDbEnhancedAsyncClient(extensions))

--- a/dynamodb-enhanced/deployment/src/test/java/io/quarkus/amazon/dynamodb/enhanced/deployment/DynamoDbEnhancedClientAllowOverrideWithProducerTest.java
+++ b/dynamodb-enhanced/deployment/src/test/java/io/quarkus/amazon/dynamodb/enhanced/deployment/DynamoDbEnhancedClientAllowOverrideWithProducerTest.java
@@ -1,0 +1,26 @@
+package io.quarkus.amazon.dynamodb.enhanced.deployment;
+
+import jakarta.inject.Inject;
+
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.test.QuarkusUnitTest;
+import software.amazon.awssdk.enhanced.dynamodb.DynamoDbEnhancedClient;
+
+public class DynamoDbEnhancedClientAllowOverrideWithProducerTest {
+
+    @Inject
+    DynamoDbEnhancedClient client;
+
+    @RegisterExtension
+    static final QuarkusUnitTest config = new QuarkusUnitTest()
+            .withApplicationRoot((jar) -> jar
+                    .addClass(DynamoDbEnhancedClientProducer.class)
+                    .addAsResource("full-config.properties", "application.properties"));
+
+    @Test
+    public void test() {
+        // should not fail
+    }
+}

--- a/dynamodb-enhanced/deployment/src/test/java/io/quarkus/amazon/dynamodb/enhanced/deployment/DynamoDbEnhancedClientProducer.java
+++ b/dynamodb-enhanced/deployment/src/test/java/io/quarkus/amazon/dynamodb/enhanced/deployment/DynamoDbEnhancedClientProducer.java
@@ -1,0 +1,24 @@
+package io.quarkus.amazon.dynamodb.enhanced.deployment;
+
+import jakarta.enterprise.context.ApplicationScoped;
+import jakarta.enterprise.inject.Produces;
+
+import software.amazon.awssdk.enhanced.dynamodb.DynamoDbEnhancedClient;
+import software.amazon.awssdk.services.dynamodb.DynamoDbClient;
+
+public class DynamoDbEnhancedClientProducer {
+
+    @Produces
+    @ApplicationScoped
+    public DynamoDbEnhancedClient enhancedClient(DynamoDbClient client) {
+        return DynamoDbEnhancedClient.builder()
+                .dynamoDbClient(client)
+                .build();
+    }
+
+    @Produces
+    @ApplicationScoped
+    public DynamoDbClient client() {
+        return DynamoDbClient.create();
+    }
+}

